Weapon effect port + various bugfixing

gamrguy opened the pull request:
Bugfixes:
- Effects didn't call ApplyItem on load
- (Rework) UpdateEquips in ModPlayer exchanged for UpdateInventory and UpdateEquip in GlobalItem
- Modifiers don't roll a null ActiveEffects if there are no available effects in the pool

Added feature: HoldItem context/hook
Added feature: hasRolled bool; set to true after first successful roll of a modifier (didn't give null). Not currently saved, so items that rolled but didn't get any effects can reroll after reloading the character.

Weapon effects have been set to have the same values as from the original mod. Calculations have been borrowed from there as well.

Some of the "debug" code has been removed, as it has been replaced with actual content.

gamrguy committed:
All of the weapon effects + bugfixes

gamrguy committed:
*whistling*

gamrguy committed:
Removed test comments

Jofairden commented on EMMItem.cs:
```diff
@@ -47,6 +119,14 @@ public override GlobalItem Clone(Item item, Item itemClone)
 			GetItemInfo(item).Modifier = Modifier._Load(tag);
+
+			// Apply on load. DUH.
+			ModifierContext ctx = new ModifierContext
```
What you're doing here is great, however, I don't think it's great applying the effects with the OnPickup context. It might be better if we'd store the passed contexts when they were being applied, which could be saved and loaded in _Load and _Save, and then could be passed here. We could still use ApplyItem, the difference being we'd have the contexts stored when they were being applied to the item.

gamrguy replied:
I understand the bit about sending OnPickup, though I could simply not set a context method at all (only sending Item). About the other part... I don't get it. ApplyItem only runs once: either when the item gets modifiers or the item loads, and for good reason, as running these effects multiple times would only distort the item's current stats beyond what the effects should be. What effect would care about whether the item was rolled or loading?

When an item rerolls (once that's implemented) it will have to SetDefaults to get its stats back to normal, then use ApplyItem again.

...of course, if we just send an Item through, then what's the point of the context in this case?

Jofairden replied:
Modders may influence behaviour of modifiers to their own desire based on present context at that time. It is a use case that must be considered, therefore we should track them as noted in my previous comment. Just in regards to not understanding the other 'part': they're not parts. It's one text and meant as one thing. We shouldn't pass any other context than ones we had stored (i.e. we should pass the stored context, rather than OnPickup) Just as a note, we're not actually storing any contexts just yet, just so you're not getting confused as to where this code may be: we should make it.

gamrguy replied:
I still don't get it. At all. We're gonna need to find some time to talk about this, as I am still under the belief that we may as well remove "context" from ApplyItem completely, as there is only one applicable context.

Jofairden commented on EMMItem.cs:
```diff
@@ -21,11 +22,46 @@ public sealed class EMMItem : GlobalItem
 		public Modifier Modifier;
 
+		// Special item parameters
+		public bool hasRolled;             // whether this item has rolled a modifier TODO: Save this
```
hasRolled is great and should stay, it however is not being saved and loaded yet. The bool should be added to _Load and _Save code.
However, effects that may influence an item do not belong in EMMItem.
Anything that influenced a rolled item should go in a separate GlobalItem to keep the code clean.
Feel free to make a new GlobalItem, it should be named ModifierItem.

Jofairden commented on EMMItem.cs:
```diff
@@ -84,11 +164,13 @@ public override bool OnPickup(Item item, Player player)
 			Modifier m = GetModifier(item);
-			if (m == null)
-				GetItemInfo(item)?.RollNewModifier(ctx);
+			if (!hasRolled && m == null)
```
I see you've incorporated hasRolled here in OnPickup, but is it also present for OnCraft?

gamrguy replied:
There shouldn't be a need for it to be in OnCraft, as it triggers when the item is crafted. Therefore every time OnCraft is called it should be on a new item with hasRolled set to false.

Similarly, it shouldn't be in PostReforge because reforging should allow items without a modifier to gain one.

Jofairden replied:
Good observation, however it's still worthwhile to add. Modders can pull finicky code and still trigger something you may not have considered (i.e. OnCraft) Consider it safeguarding, it might not happen in the regular use case but it may in another you hadn't conceived. Doing such safeguarding is best practice in my view, so we should do it.

Jofairden replied:
On another note, we will need to add additional checks in save and load, because now we've included hasRolled and it can save if it has rolled. In _Save we must check if Modifier is not null, if it is null we should put the Modifier as null in the TC. (otherwise we do regular saving) In _Load, we must check if the Modifier key in TC is null, if not then we can load it.

Jofairden commented on EMMPlayer.cs:
```diff
@@ -13,29 +14,52 @@ public class EMMPlayer : ModPlayer
 
-		public override void UpdateEquips(ref bool wallSpeedBuff, ref bool tileSpeedBuff, ref bool tileRangeBuff)
+		// Globals for modifiers
+		public bool holdingCursed;    // Whether currently holding a cursed item (take 1 damage per second)
```
Same feedback as for EMMItem, however, there isn't anything generic right now for EMMPlayer. 
Feel free to make a new ModPlayer and name it ModifierPlayer, and leave EMMPlayer empty for now or remove it.

Jofairden commented on Effects/WeaponEffect.cs:
```diff
@@ -0,0 +1,331 @@
+
+namespace Loot.Effects.WeaponEffects
+{
+	public class DamagePlusEffect : ModifierEffect
```
Though I understand ease of access for effects to be in a single file, it is favorable for them to be in separate classes. This keeps the size per file fairly small, and excludes code from other effects that may not be of interest.

Jofairden reviewed (changes requested):
This is looking decent. I've left a few comments on things that should be addressed

gamrguy committed:
Add hasRolled checks to OnCraft and OnReforge

gamrguy commented:
Thanks for the review! I'll work on these requests.

gamrguy committed:
removed hasRolled checks

gamrguy committed:
Code cleanup + hasRolled saving

Separated parts of EMMItem into ModifierItem.
EMMPlayer renamed to ModifierPlayer.
hasRolled now saved and loaded.

gamrguy committed:
Fix hasRolled saving

gamrguy committed:
Effect class separation

Separated the various Effect classes into their own files

gamrguy commented:
Alright, how does this look?

Jofairden reviewed (changes requested):
Updated discussions. 
Since it didn't really pull through the comments very well, I'll link them:
[Here](https://github.com/Jofairden/EvenMoreModifiers/pull/3#discussion_r175931811) and [here](https://github.com/Jofairden/EvenMoreModifiers/pull/3#discussion_r175930068)

For now, you disregard the way we deal with contexts (the way  right now we have HoldItem, UpdateItem and so forth). I'm having further internal discussion regarding this, I'll get back to you about it in this PR.

Jofairden commented on EMMItem.cs:
```diff
@@ -119,19 +89,20 @@ public override GlobalItem Clone(Item item, Item itemClone)
 		{
-			return Modifier.Save(GetItemInfo(item).Modifier);
+			var tag = Modifier.Save(GetItemInfo(item).Modifier);
+			tag.Add("HasRolled", hasRolled);
```
HasRolled should be saved in Modifier.Save, not here. See one of my earlier comments. Since it will need access to both Modifier and hasRolled, I suggest we change the static Save so it accepts the item info object instead. (this should also make it easier for future saving changes)

e: It looks like github is quite buggy on my end, I'm seeing my comments appear all over the place. I'll end commenting now, lol.

gamrguy commented:
I already handled that mess with additional checks. Did you even look at the changes in d8d4b1f (Fix hasRolled saving)? NeedsSaving was edited to accept items that were rolled but did not have a modifier. Save and Load check for if the item has a modifier or if the saved data contains a modifier, respectively, and ignores Modifier.Save/Load respectively, as we're *not saving or loading modifiers here*.

Also, Modifier.Save/Load etc. does not refer back to the item! What you're suggesting is impossible. It also doesn't make sense, as `hasRolled` is a property of the *item*, not of its modifier.

Jofairden commented:
Yes I did see the changes, I forgot to edit / delete my comment (saw them later, I missed them in the changes panel) In regards to the saving loading checks: 👍 However, I think you're slightly missing an overarching concept here. Right now you've added those checks, and saving hasRolled, to the non static overload of Save. The concept here is that modders can choose override saving and loading. The idea is that the static method defines generic (default) saving for an EMMItem (an item in the context of 'it can have a modifier, and it may or may not have rolled yet'). This means by default, the overload of Save should return what the static Save does. So to put it simply, the checks and saving hasRolled need to be put inside the static Save method.  This is why I suggested we pass the entire EMMItem instance to the static save method rather than only the modifier. I also recommend we still save a key 'Modifier' rather than an empty TC, in my opinion it's clearer and likely less error-prone since the key will always be present. Null-checking / Null-coalescing is always a thing.

Consider a modder decides to do custom saving. They will most likely still want default saving, so they'd call the static Save version. However, it does not include hasRolled, as you implemented it in the overloaded version. I hope now it all makes a bit more sense.

gamrguy commented:
But how would this "custom saving" manage to override the save/load methods in EMMItem? That doesn't make sense to me; every item has its own instance of EMMItem, and each EMMItem has the exact same non-static methods. Do you have some kind of back-end shenanigans going on that would actually manage to disable this?

Jofairden commented:
So... just leave it as is. I'll do it some other time

gamrguy committed:
hasRolled check in OnCraft
